pkg/controller/replica.go

```diff
@@ -0,0 +1,335 @@
+			// underlying volume does not exist, so volume attachment cannot be made
+			return nil
+		}
+		if err = r.createReplicas(ctx, min(defaultMaxReplicaUpdateCount, desiredReplicaCount-currentReplicaCount), azVolume.Name, azVolume.Status.Detail.ResponseObject.VolumeID); err != nil {
+			klog.Errorf("failed to create %d replicas for volume (%s): %v", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileReplica) getNodesForReplica(ctx context.Context, volumeName string, numReplica int) ([]string, error) {
+	pods, err := getPodsFromVolume(ctx, r, volumeName)
+	if err != nil {
+		return nil, err
+	}
+	nodes, err := getNodesForReplica(ctx, r, nil, pods...)
+	if err != nil {
+		return nil, err
+	}
+
+	replicaNodes, err := getNodesWithReplica(ctx, r, volumeName)
+	if err != nil {
+		return nil, err
+	}
+
+	skipSet := map[string]bool{}
+	for _, replicaNode := range replicaNodes {
+		skipSet[replicaNode] = true
+	}
+
+	filtered := []string{}
+	numFiltered := 0
+	for _, node := range nodes {
+		if numFiltered >= numReplica {
+			break
+		}
+		if skipSet[node] {
+			continue
+		}
+		filtered = append(filtered, node)
+		numFiltered++
+	}
+
+	return filtered, nil
+}
+
+func (r *ReconcileReplica) createReplicas(ctx context.Context, numReplica int, volumeName, volumeID string) error {
+	// if volume is scheduled for clean up, skip replica creation
+	if _, cleanUpScheduled := r.cleanUpMap.get(volumeName); cleanUpScheduled {
+		return nil
+	}
+
+	nodes, err := r.getNodesForReplica(ctx, volumeName, numReplica)
+	if err != nil {
+		klog.Errorf("failed to get a list of nodes for replica attachment: %v", err)
+		return err
+	}
+
+	for _, node := range nodes {
+		if err := createReplica(ctx, r, volumeID, node); err != nil {
+			klog.Errorf("failed to create replica azVolumeAttachment for volume %s: %v", volumeName, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func NewReplicaController(mgr manager.Manager, azVolumeClient azClientSet.Interface, kubeClient kubeClientSet.Interface, podToClaimsMap, claimToPodsMap, claimToVolumeMap, volumeToClaimMap *SharedMap) (*ReconcilePod, error) {
+	logger := mgr.GetLogger().WithValues("controller", "replica")
+	reconciler := ReconcilePod{
+		client:           mgr.GetClient(),
+		azVolumeClient:   azVolumeClient,
+		kubeClient:       kubeClient,
+		podToClaimsMap:   podToClaimsMap,
+		claimToPodsMap:   claimToPodsMap,
+		claimToVolumeMap: claimToVolumeMap,
+		volumeToClaimMap: volumeToClaimMap,
+	}
+	c, err := controller.New("replica-controller", mgr, controller.Options{
+		MaxConcurrentReconciles: 10,
+		Reconciler:              &reconciler,
+		Log:                     logger,
+	})
+
+	if err != nil {
+		klog.Errorf("Failed to create replica controller. Error: (%v)", err)
+		return nil, err
+	}
+
+	klog.V(2).Info("Starting to watch Pod.")
```

Comment:
Starting to watch AzVolumeAttachment objects? The comment below should also be fixed.